waspc/packages/deploy/src/providers/railway/commands/setup/setup.ts

```diff
@@ -0,0 +1,197 @@
+    waspSays("Postgres service already exists. Skipping database setup.");
+  } else {
+    await setupDb(deploymentInstructions);
+  }
+
+  if (project.doesServiceExist(deploymentInstructions.clientServiceName)) {
+    waspSays("Client service already exists. Skipping client setup.");
+  } else {
+    await setupClient(deploymentInstructions);
+  }
+
+  if (project.doesServiceExist(deploymentInstructions.serverServiceName)) {
+    waspSays("Server service already exists. Skipping server setup.");
+  } else {
+    await setupServer(deploymentInstructions);
+  }
+}
+
+async function setupRailwayProjectForDirectory({
+  railwayExe,
+  projectName,
+  waspProjectDir,
+  existingProjectId,
+}: {
+  railwayExe: RailwayCliExe;
+  projectName: RailwayProjectName;
+  waspProjectDir: WaspProjectDir;
+  existingProjectId: RailwayProjectId | null;
+}): Promise<RailwayProject> {
+  const { status, project } = await getRailwayProjectStatus({
+    projectName,
+    waspProjectDir,
+    railwayExe,
+    existingProjectId,
+  });
+
+  switch (status) {
+    case ProjectStatus.EXISTING_PROJECT_ALREADY_LINKED:
+      waspSays(
+        `Project with name "${projectName}" already linked. Skipping project creation.`,
+      );
+      return project;
+
+    case ProjectStatus.EXISTING_PROJECT_SHOULD_BE_LINKED:
+      waspSays(`Linking project with name "${project.name}" to this directory`);
+      return linkRailwayProjectToWaspProjectDir(project, {
+        railwayExe,
+        waspProjectDir,
+      });
+
+    case ProjectStatus.MISSING_PROJECT:
+      waspSays(`Setting up Railway project with name "${projectName}"`);
+      return initRailwayProject({
+        projectName,
+        railwayExe,
+        waspProjectDir,
+      });
+
+    default:
+      status satisfies never;
+      throw new Error(`Unhandled status: ${status}`);
+  }
+}
+
+async function setupDb({
+  cmdOptions: options,
+}: DeploymentInstructions<SetupCmdOptions>): Promise<void> {
+  waspSays("Setting up database");
+
+  const railwayCli = createCommandWithCwd(
+    options.railwayExe,
+    options.waspProjectDir,
+  );
+  await railwayCli(["add", "-d", "postgres"]);
+}
+
+async function setupServer({
+  cmdOptions: options,
+  serverServiceName,
+  clientServiceName,
+  dbServiceName,
+}: DeploymentInstructions<SetupCmdOptions>): Promise<void> {
+  waspSays(`Setting up server app with name ${serverServiceName}`);
+
+  // The client service needs a URL so it can be referenced in the
+  // server service env variables.
+  await generateServiceUrl(clientServiceName, clientAppPort, options);
+
+  const serverBuildDir = getServerBuildDir(options.waspProjectDir);
+  const railwayCli = createCommandWithCwd(options.railwayExe, serverBuildDir);
+
+  const clientUrl = getRailwayPublicUrlEnvVarForService(clientServiceName);
+  const databaseUrl = getRailwayDatabaseUrlEnvVar(dbServiceName);
+  const serverUrl = getRailwayPublicUrlEnvVarForSelf();
```

Comment:
I think I commented somewhere else, but I was confused by `forSelf` -> how does it know that Self is server in this context? There is some "assumed state", to put it that way?